src/material_library.cpp

```diff
@@ -0,0 +1,376 @@
+    pyne::Material mat = Material(mat_lib.get_material(*it));
+    (*this).add_material(*it, mat);
+  }
+}
+
+void pyne::MaterialLibrary::merge(pyne::MaterialLibrary* mat_lib) {
+  merge(*mat_lib);
+}
+
+void pyne::MaterialLibrary::load_json(Json::Value json) {
+  Json::Value::Members keys = json.getMemberNames();
+  Json::Value::Members::const_iterator ikey = keys.begin();
+  Json::Value::Members::const_iterator ikey_end = keys.end();
+  for (; ikey != ikey_end; ++ikey) {
+    pyne::Material mat = pyne::Material();
+    mat.load_json(json[*ikey]);
+    (*this).add_material(*ikey, Material(mat));
+  }
+}
+
+Json::Value pyne::MaterialLibrary::dump_json() {
+  Json::Value json = Json::Value(Json::objectValue);
+
+  for (auto name : keylist) {
+    json[name] = material_library[name]->dump_json();
+  }
+  return json;
+}
+
+void pyne::MaterialLibrary::from_json(const std::string& filename) {
+  if (!pyne::file_exists(filename)) throw pyne::FileNotFound(filename);
+  std::string s;
+  std::ifstream f(filename.c_str(), std::ios::in | std::ios::binary);
+  f.seekg(0, std::ios::end);
+  s.resize(f.tellg());
+  f.seekg(0, std::ios::beg);
+  f.read(&s[0], s.size());
+  f.close();
+  Json::Reader reader;
+  Json::Value json;
+  reader.parse(s, json);
+  load_json(json);
+}
+
+void pyne::MaterialLibrary::write_json(const std::string& filename) {
+  Json::Value json = dump_json();
+  Json::StyledWriter writer;
+  std::string s = writer.write(json);
+  std::ofstream f(filename.c_str(), std::ios_base::trunc);
+  f << s << "\n";
+  f.close();
+}
+
+int pyne::MaterialLibrary::ensure_material_number(pyne::Material& mat) const {
+  int mat_number = -1;
+  if (mat.metadata.isMember("mat_number")) {
+    if (Json::intValue <= mat.metadata["mat_number"].type() &&
+        mat.metadata["mat_number"].type() <= Json::realValue) {
+      mat_number = mat.metadata["mat_number"].asInt();
+    } else {
+      mat_number = std::stoi(mat.metadata["mat_number"].asString());
+      mat.metadata["mat_number"] = mat_number;
+    }
+    std::set<int>::iterator mat_numb_it;
+    mat_numb_it = mat_number_set.find(mat_number);
+    if (mat_numb_it != mat_number_set.end()) {
+      std::string msg = "Material number ";
+      msg += mat_number;
+      msg += " is already in the library.";
+      warning(msg);
+    }
+  }
+  return mat_number;
+}
+
+std::string pyne::MaterialLibrary::ensure_material_name(
+    pyne::Material& mat) const {
+  std::string mat_name = "";
+  int mat_number = ensure_material_number(mat);
+
+  if (mat.metadata.isMember("name")) {
+    if (Json::intValue <= mat.metadata["name"].type() &&
+        mat.metadata["name"].type() <= Json::realValue) {
+      mat_name = std::to_string(mat.metadata["name"].asInt());
+      mat.metadata["name"] = mat_name;
+    } else {
+      mat_name = mat.metadata["name"].asString();
+    }
+  } else {
+    if (mat_number == -1) {
+      mat_number = keylist.size();  // set a temp mat_number to form mat name
```

Comment:
Should we set the metadata for `mat_number` as well?

Reply:
I now remember... sorry it took me so long....

we use this number to name the material that has neither name or mat_number, but the number was not store in the previous API.
we can choose to define name and/or mat_number metadata or neither (just defining a key), I don't know what is best.

Reply:
Now that we are using the maximum entry in `keylist` to make up a number, it seems that we should add that number to `keylist` so that it's correct for the next time.  Therefore, it also makes sense to assign `mat_number` metadata, I think.